library/swift/RequestBuilder.swift

```diff
@@ -0,0 +1,106 @@
+import Foundation
+
+/// Builder used for constructing instances of `Request` types.
+@objcMembers
+public final class RequestBuilder: NSObject {
+  /// URL for the request.
+  public private(set) var url: URL
+  /// Method for the request.
+  public private(set) var method: RequestMethod
+  /// Headers to send with the request.
+  /// Multiple values for a given name are valid, and will be sent as comma-separated values.
+  public private(set) var headers: [String: [String]] = [:]
+  /// Trailers to send with the request.
+  /// Multiple values for a given name are valid, and will be sent as comma-separated values.
+  public private(set) var trailers: [String: [String]] = [:]
+  // Serialized data to send as the body of the request.
+  public private(set) var body: Data?
+  // Retry policy to use for this request.
+  public private(set) var retryPolicy: RetryPolicy?
+
+  // MARK: - Initializers
+
+  /// Internal initializer used for converting a request back to a builder.
+  init(request: Request) {
+    self.url = request.url
+    self.method = request.method
+    self.headers = request.headers
+    self.trailers = request.trailers
+    self.body = request.body
+    self.retryPolicy = request.retryPolicy
+  }
+
+  /// Public initializer.
+  public init(url: URL, method: RequestMethod) {
+    self.url = url
+    self.method = method
+  }
+
+  // MARK: - Builder functions
+
+  @discardableResult
+  public func addHeader(name: String, value: String) -> RequestBuilder {
+    self.headers[name, default: []].append(value)
+    return self
+  }
+
+  @discardableResult
+  public func removeHeader(name: String) -> RequestBuilder {
+    self.headers.removeValue(forKey: name)
```

Comment:
nit: this might lead to confusion since it will remove _all_ headers with a matching name. (e.g. `Link` headers). We might want to rename this to better indicate that, and/or provide a pair-matching API for header removal.

Reply:
That's a good callout. What about:

```swift
/// Remove all headers with this name
func removeHeaders(named name: String)

/// Remove an exact match header
func removeHeader(name: String, value: String)
```

Reply:
Yep, I like that.

Reply:
Done